NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,344 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+		FirstHandle( first ),
+		ResultHandle( result ),
+		MultiplierHandle( multiplier )
+	{}
+
+	void Run( const CKMeansClustering&, int /*threadIndex*/, int index, int count ) override
+	{ MathEngine.VectorMultiply( FirstHandle + index, ResultHandle + index, count, MultiplierHandle ); }
+
+	IMathEngine& MathEngine;
+	const CConstFloatHandle& FirstHandle;
+	const CFloatHandle& ResultHandle;
+	const CConstFloatHandle& MultiplierHandle;
+};
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CMatrixMultiplyArgs final : public CKMeansClustering::IArgs {
+	CMatrixMultiplyArgs( int batchSize, IMathEngine& mathEngine,
+			const CConstFloatHandle& first, int firstHeight, int firstWidth,
+			const CConstFloatHandle& second, int secondHeight,
+			const CFloatHandle& result, int resultBufferSize ) :
+		IArgs( batchSize ),
+		MathEngine( mathEngine ),
+		FirstHandle( first ),
+		FirstHeight( firstHeight ),
+		FirstWidth( firstWidth ),
+		SecondHandle( second ),
+		SecondHeight( secondHeight ),
+		ResultHandle( result ),
+		ResultBufferSize( resultBufferSize )
+	{}
+
+	void Run( const CKMeansClustering&, int, int, int ) override { NeoPresume( false ); }
+	void Run( const CKMeansClustering&, int, int, int, int, int );
+
+	IMathEngine& MathEngine;
+	const CConstFloatHandle& FirstHandle;
+	const int FirstHeight;
+	const int FirstWidth;
+	const CConstFloatHandle& SecondHandle;
+	const int SecondHeight;
+	const CFloatHandle& ResultHandle;
+	const int ResultBufferSize;
+};
+
+void CKMeansClustering::CMatrixMultiplyArgs::Run( const CKMeansClustering&, int /*threadIndex*/,
+	int firstHeightStart, int firstHeightCount, int secondHeightStart, int secondHeightCount )
+{
+	const CConstFloatHandle& firstData = FirstHandle + firstHeightStart * FirstWidth;
+	const CConstFloatHandle& const secondData = SecondHandle + secondHeightStart * FirstWidth;
+	const CFloatHandle& resultData = ResultHandle + firstHeightStart * SecondHeight + secondHeightStart;
+
+	MathEngine.MultiplyMatrixByTransposedMatrix( /*batch*/Size(),
+		firstData, firstHeightCount, FirstWidth,
+		secondData, secondHeightCount, resultData, ResultBufferSize );
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CThreadTask final {
+	CThreadTask( const CKMeansClustering& owner, IArgs& args ) :
+		owner( owner ),
+		args( args )
+	{}
+
+	void CallRun( int threadIndex, int align = 1 );
+	void CallRunMxMT( int threadIndex );
+
+private:
+	const CKMeansClustering& owner;
+	IArgs& args;
+};
+
+void CKMeansClustering::CThreadTask::CallRun( int threadIndex, int align )
+{
+	int index = 0;
+	int count = 0;
+	if( GetTaskIndexAndCount( owner.params.ThreadCount, threadIndex, args.Size(), align, index, count ) ) {
+		args.Run( owner, threadIndex, index, count );
+	}
+}
+
+void CKMeansClustering::CThreadTask::CallRunMxMT( int threadIndex )
```

Comment:
Isn't it better to just leave everything related to matrix multiplication outside of interfaces?
The `CheckCast` here with `NeoPresume( false )` in `CMatrixMultiplyArgs::Run` show that it should be outside of this interface.
My idea: `CMatrixMultiplyArgs` shouldn't inherit `IArgs` and should contain only data.
This launching and indexing should be inside of lambda-function which is passed to `NEOML_NUM_THREADS` in `matrixMultiplyByTransposed`

Reply:
I think we will rename it and move it right here to KMeans (during the final stage).